swift/Sources/Lexer/Lexer.swift

```diff
@@ -9,72 +6,55 @@ extension Character {
 
 extension String {
     func numberString(from index: String.Index) -> String {
-
         var endIndex = self.index(after: index)
-
-        while endIndex < self.endIndex {
-            if self[endIndex].isNumber == false {
-                return String(self[index..<endIndex])
-            }
+        while endIndex < self.endIndex && self[endIndex].isNumber {
             endIndex = self.index(after: endIndex)
         }
-        return String(self[index...])
+        return String(self[index..<endIndex])
     }
 
     func identifier(from index: String.Index) -> String {
         var endIndex = self.index(after: index)
-
-        while endIndex < self.endIndex {
-            if self[endIndex].isValidIdentifier == false {
-                return String(self[index..<endIndex])
-            }
+        while endIndex < self.endIndex && self[endIndex].isValidIdentifier {
             endIndex = self.index(after: endIndex)
         }
-        return String(self[index...])
+        return String(self[index..<endIndex])
     }
 }
 
 struct Lexer {
 
     let input: String
-    var currentIndex: String.Index
-
-    init(input: String) {
-        self.input = input
-        self.currentIndex = input.startIndex
+    lazy var currentIndex = input.startIndex
+    var currentChar: Character {
+        mutating get {
```

Comment:
We never set the `currentChar`, so we can change this to:
`var currentChar: Character { input[currentIndex] }`

Reply:
Due to `Lexer` being a struct, each of its properties are immutable - even if marked as `var`.  In order to change them, a function must be marked as `mutable`.

In this case the function is the getter.  I tried removing `mutating` again just to check, but I am still getting the following error:

`cannot use mutating getter on immutable value: 'self' is immutable`

Are you not getting this error when you try to compile?

Thank you again for considering my PR!